### ISSUE TITLE
vterm-line-mode

### ISSUE THREAD

**Issue by guskovd** [akermu/emacs-libvterm#75]
It would be great to implement vterm-line-mode, just as it was done in term / ansi-term.
This is very useful when searching in the output terminal, copying text, etc.

**Comment by QiangF** [comment]
Vterm-mode works like a text buffer with Evil-mode, I think you can do all you want already.

**Comment by Sbozzolo** [comment]
If the purpose of the line-mode is for searching and copying text, wouldn't be better if vterm supported these features directly (ie without having to enable a line-mode)?

**Comment by jixiuf** [comment]
keybindings like `C-n` `C-p` `C-f` `C-b` behave differently in char-mode and line-mode ,I think it is not easy  to support it directly

**Comment by Sbozzolo** [comment]
Yes, you are right. However, if I look at the way I work, if I could
mark and search directly in char-mode I would have no need to go to
line-mode. For my personal workflow, I would like C-p, C-f, C-b, C-n to
always move the point. Then, if want to recall the last command I gave
to the shell I would use M-p or the arrows.

For what reasons do people use the line-mode, a part from searching and
marking? (This is not a rhetorical question)


jixiuf writes:

**Comment by guskovd** [comment]
> Yes, you are right. However, if I look at the way I work, if I could mark and search directly in char-mode I would have no need to go to line-mode. For my personal workflow, I would like C-p, C-f, C-b, C-n to always move the point. Then, if want to recall the last command I gave to the shell I would use M-p or the arrows. For what reasons do people use the line-mode, a part from searching and marking? (This is not a rhetorical question) jixiuf writes:
> […](#)
> keybindings like C-n C-p C-f C-b behave differently in char-mode and line-mode ,I think it is not easy to support it directly — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or mute the thread.*

In term, I mainly use line-mode. It is quite convenient not only for text search and navigation.
Using a single history for multiple sessions (including ssh) - killing feature for me

**Comment by apauzies** [comment]
Me too, can we merge @jixiuf change?
Work fine for me

**Comment by jixiuf** [comment]
don't known whether this change is necessary.
@akermu what's your opinions

**Comment by cullan** [comment]
I love vterm and not being able to copy text is the only thing making me need to run a separate terminal occasionally.

**Comment by Sbozzolo** [comment]
I use the following (hacky) trick. I defined a function that recognizes if the point is in the input line or not. If it is not, then all the commands behave like a normal text buffer, otherwise they behave like a `vterm` buffer. This allows me to copy the output of shell commands. This solution is far from perfect because to find whether I am on the input line or not I look at the first characters of the line, if they match a known prompt, then I declare that that is an input line. For instance, my zsh prompt starts with `$ `, so I look at the first two characters, and they are `$ `, then the commands I execute have to be treated differently. Similarly with `gnuplot` and the prompt `gnuplot>`, and so on. This is hacky way to deal with the problem and it works 90% of the time. However, it was crucial in me moving to `vterm` as a daily driver.

The code is the following:
```emacs-lisp
  (defun starts-with-p (line prompt)
    "Check if LINE starts with PROMPT"
    (string= (substring line 0 (length prompt)) prompt)
    )

  (defun is-a-prompt-p (line)
    "Check if the LINE is the input line of some interactive process."
    (or
     (starts-with-p line "$")
     (starts-with-p line "gnuplot>")
     (starts-with-p line ">>>")
     )
    )

  (defun send-command-input (command &optional args)
    "Check if the point is in the input line, if it is, send the
     command with vterm. This works only if the first characters
     of the line are in the database of is-a-prompt-p."
    (if (is-a-prompt-p (thing-at-point 'line))
        (vterm--self-insert)            ; Prompt line
      (funcall command args)
      )
    )

  (define-key vterm-mode-map (kbd "C-f") '(lambda () (interactive) (send-command-input 'forward-char)))
  (define-key vterm-mode-map (kbd "C-b") '(lambda () (interactive) (send-command-input 'backward-char)))
  (define-key vterm-mode-map (kbd "C-n") nil)
  (define-key vterm-mode-map (kbd "C-p") nil)
  (define-key vterm-mode-map (kbd "C-y") 'vterm-yank)
  (define-key vterm-mode-map (kbd "<mouse-2>") 'vterm-yank)
  (define-key vterm-mode-map (kbd "C-s") '(lambda () (interactive) (send-command-input 'counsel-grep-or-swiper)))
  (define-key vterm-mode-map (kbd "C-r") '(lambda () (interactive) (send-command-input 'counsel-grep-or-swiper)))
  (define-key vterm-mode-map (kbd "M-f") '(lambda () (interactive) (send-command-input 'forward-word)))
  (define-key vterm-mode-map (kbd "M-b") '(lambda () (interactive) (send-command-input 'backward-word)))
  (define-key vterm-mode-map (kbd "C-e") '(lambda () (interactive) (send-command-input 'move-end-of-line)))
  (define-key vterm-mode-map (kbd "C-a") '(lambda () (interactive) (send-command-input 'move-beginning-of-line)))
  (define-key vterm-mode-map (kbd "M-w") '(lambda () (interactive) (send-command-input
                                                               '(lambda (interactive)
                                                                  (kill-ring-save (point) (mark t))
                                                                  ))))

  (define-key vterm-mode-map (kbd "C-SPC") '(lambda () (interactive) (send-command-input 'set-mark-command)))
```

**Comment by akermu** [comment]
My idea would be to have a a `vterm-copy-mode` minor mode, which behaves similar to tmux copy mode. In essence during this mode, the buffer is a normal `read-only` text buffer, so every keybinding should works as usual. 

During this time rendering is blocked and after exiting the minor mode the whole buffer is rendered again, so we are in a consistent state. Another solution would be to buffer terminal output until the minor mode is disabled and process it afterwards, but the buffer could grow infinitely (you enter the copy mode during a htop session and never leave it).

**Comment by jixiuf** [comment]
maybe we can freeze the terminal by send a `ctrl-s` when enable `vterm-copy-mode`
and unfreeze it by sending `ctrl-q` after disable `vterm-copy-mode'
https://unix.stackexchange.com/questions/137842/what-is-the-point-of-ctrl-s

**Comment by akermu** [comment]
This should be solved with `vterm-copy-mode' integrated in #108.
